lib/charms/haproxy/v0/haproxy_route.py

```diff
@@ -0,0 +1,1191 @@
+    )
+
+
+class HaproxyRouteProviderAppData(_DatabagModel):
+    """haproxy-route provider databag schema.
+
+    Attributes:
+        endpoints: The list of proxied endpoints that maps to the backend.
+    """
+
+    endpoints: list[AnyHttpUrl]
+
+
+class RequirerUnitData(_DatabagModel):
+    """haproxy-route requirer unit data.
+
+    Attributes:
+        host: hostname or IP address of the unit.
+    """
+
+    host: str = Field(description="Hostname or IP address of the unit.")
+
+
+@dataclass
+class HaproxyRouteRequirerData:
+    """haproxy-route requirer data.
+
+    Attributes:
+        application_data: Application data.
+        units_data: Units data
+    """
+
+    application_data: RequirerApplicationData
+    units_data: list[RequirerUnitData]
+
+
+@dataclass
+class HaproxyRouteRequirersData:
+    """haproxy-route requirers data.
+
+    Attributes:
+        requirers_data: List of requirer data.
+    """
+
+    requirers_data: list[HaproxyRouteRequirerData]
+
+
+class HaproxyRouteEnpointsAvailableEvent(EventBase):
+    """HaproxyRouteEnpointsAvailableEvent custom event.
+
+    This event indicates that the haproxy-route endpoints are available.
+    """
+
+
+class HaproxyRouteProviderEvents(CharmEvents):
+    """List of events that the TLS Certificates requirer charm can leverage.
+
+    Attributes:
+        endpoints_available: This event indicates that
+            the haproxy-route endpoints are available.
+    """
+
+    endpoints_available = EventSource(HaproxyRouteEnpointsAvailableEvent)
+
+
+class HaproxyRouteProvider(Object):
+    """Haproxy-route interface provider implementation.
+
+    Attributes:
+        on: Custom events of the provider.
+        relations: Related appliations.
+    """
+
+    on = HaproxyRouteProviderEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = HAPROXY_ROUTE_RELATION_NAME,
+    ) -> None:
+        """Initialize the HaproxyRouteProvider.
+
+        Args:
+            charm: The charm that is instantiating the library.
+            relation_name: The name of the relation.
+        """
+        super().__init__(charm, relation_name)
+
+        self._relation_name = relation_name
+        self.charm = charm
+        on = self.charm.on
+        self.framework.observe(on[self._relation_name].relation_created, self._configure)
+        self.framework.observe(on[self._relation_name].relation_changed, self._configure)
+        self.framework.observe(on[self._relation_name].relation_broken, self._configure)
+
+    @property
+    def relations(self) -> list[Relation]:
+        """The list of Relation instances associated with this endpoint."""
+        return list(self.charm.model.relations[self._relation_name])
+
+    def _configure(self, _: EventBase) -> None:
+        """Handle relation events."""
+        try:
+            if relations := self.relations:
+                self.get_data(relations)
```

Comment:
We should probably use `_ = self.get_data(relations)`

Reply:
updated